torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1719,8 +1744,60 @@ def _update_divisible(self):
 
         self.divisible = new_divisible
 
+    @_lru_cache
+    def maybe_simplify_floordiv(self, expr: "sympy.Expr") -> "sympy.Expr":
+        # a // b == expr
+        # => a >= (b * expr) and a < ((b + 1) * expr)
+        if isinstance(expr, sympy.Eq) and isinstance(expr.lhs, FloorDiv):
+            numerator, denominator = expr.lhs.args
+            expr = sympy.And(
+                sympy.Ge(numerator, (expr.rhs * denominator)),  # type: ignore
+                sympy.Lt(numerator, ((expr.rhs + 1) * denominator))  # type: ignore
+            )
+        # a // b != expr
+        # => a < (b * expr) or a >= ((b + 1) * expr)
+        if isinstance(expr, sympy.Ne) and isinstance(expr.lhs, FloorDiv):
+            numerator, denominator = expr.lhs.args
+            expr = sympy.Or(
+                sympy.Lt(numerator, (expr.rhs * denominator)),  # type: ignore
+                sympy.Ge(numerator, ((expr.rhs + 1) * denominator))  # type: ignore
+            )
+
+        # The transformations below only work if b is positive.
+        # Note: we only have this information for constants.
+        def is_floordiv_with_positive_denominator(e) -> bool:
+            if not isinstance(e, FloorDiv):
+                return False
+            number = e.args[1]
+            return isinstance(number, sympy.Integer) and bool(number > 0)
+
+        # a // b > expr  => a >= (b + 1) * expr
+        # a // b >= expr => a >= b * expr
+        if isinstance(expr, (sympy.Gt, sympy.Ge)) and is_floordiv_with_positive_denominator(expr.lhs):
+            quotient = expr.rhs if isinstance(expr, sympy.Ge) else (expr.rhs + 1)  # type: ignore
+            expr = sympy.Ge(expr.lhs.args[0], (quotient * expr.lhs.args[1]))  # type: ignore
+        # a // b < expr  => a < b * expr
+        # a // b <= expr => a < (b + 1) * expr
+        if isinstance(expr, (sympy.Lt, sympy.Le)) and is_floordiv_with_positive_denominator(expr.lhs):
+            quotient = expr.rhs if isinstance(expr, sympy.Lt) else (expr.rhs + 1)  # type: ignore
+            expr = sympy.Lt(expr.lhs.args[0], (quotient * expr.lhs.args[1]))  # type: ignore
```

Comment:
<details>
<summary>Z3 Proof</summary>

```python
tests = [
    (
        # Less-Than
        lambda a,b,c: a / b < c,
        lambda a,b,c: a < c * b,
    ),
    (
        # Less-Than or Equal
        lambda a,b,c: a / b <= c,
        lambda a,b,c: a < (c + 1) * b,
    ),
    (
        # Greater-Than
        lambda a,b,c: a / b > c,
        lambda a,b,c: a >= (c + 1) * b,
    ),
    (
        # Greater-Than or Equal
        lambda a,b,c: a / b >= c,
        lambda a,b,c: a >= c * b,
    ),
]

for old, new in tests:
    a = z3.Int("a")
    b = z3.Int("b")
    c = z3.Int("c")

    solver = z3.Solver()

    solver.add(b > 0)
    solver.add(z3.And(old(a, b, c), z3.Not(new(a, b, c))))

    if solver.check() == z3.sat:
        m = solver.model()
        print(m)
    else:
        print("unsat")
```

</details>

Reply:
Does Z3 integer division agree with the floordiv we're using here?

The semantics of `floordiv` in Z3 are given as per https://smtlib.cs.uiowa.edu/theories-Ints.shtml (and the formulas in the paper in there), by:
![image](https://user-images.githubusercontent.com/3291265/228805914-2893b209-7573-42a2-9375-c021182884fc.png)
This is pretty much an integer division defined so that `0 <= a % b < abs(b)` for every `a,b` irrespective of the sign of `a` and `b`. Note that this is different to Python's division.

Which division are we using when have floordiv in our formulas?

Reply:
Ah. You are right. As far as I understand, `FloorDiv` is rounding towards -infinity. I have added an integer division function that hopefully has the same semantics as that.

```python
def div(a, b):
    return (b < 0) * (-a / b) + (b >= 0) * (a / b)
```

Reply:
From the link you posted (+ I actually ran `model.evaluate` to check):

> Regardless of sign of m, 
  when n is positive, (div m n) is the floor of the rational number m/n;
  when n is negative, (div m n) is the ceiling of m/n.

Reply:
Ah... I missed one negative sign on `b`.